File: daily/Binary Search/410-Split-array-largest-sum.py
class Solution:
    def splitArray(self, nums: List[int], k: int) -> int:
        # Search space:
        # - Minimum possible largest sum = max element
        # - Maximum possible largest sum = sum of all elements
        left, right = max(nums), sum(nums)

        while left < right:
            mid = (left + right) // 2

            # Check feasibility: can we split into <= k subarrays
            # such that each subarray sum <= mid?
            if self.canSplit(nums, mid, k):
                right = mid      # try smaller maximum sum
            else:
                left = mid + 1   # need larger maximum sum

        return left

    def canSplit(self, nums: List[int], limit: int, k: int) -> bool:
        count = 1
        current_sum = 0

        for num in nums:
            current_sum += num
            if current_sum > limit:
                count += 1
                current_sum = num
                if count > k:
                    return False

        return True

File: daily/Binary Search/410-Split-array-largest-sum.go
package daily

func splitArray(nums []int, k int) int {
	maxValue := 0
	for _, num := range nums {
		if num > maxValue {
			maxValue = num
		}
	}

	sumArr := 0
	for _, num := range nums {
		sumArr += num
	}

	ans := 0
	lo, hi := maxValue, sumArr
	for lo <= hi {
		mid := (lo + hi) / 2
		if valueFit(nums, mid, k) {
			ans = mid
			hi = mid - 1
		} else {
			lo = mid + 1
		}
	}
	return ans
}

func valueFit(nums []int, limit int, k int) bool {
	count := 1
	sum := 0
	for _, num := range nums {
		sum += num
		if sum > limit {
			count++
			sum = num
		}
	}
	return count <= k
}
